stdlib/public/core/TemporaryAllocation.swift

```diff
@@ -0,0 +1,265 @@
+@_alwaysEmitIntoClient @_transparent
+internal func _byteCountForTemporaryAllocation<T>(
+  of type: T.Type,
+  capacity: Int
+) -> Int {
+  // PRECONDITIONS: Negatively-sized buffers obviously cannot be allocated on
+  // the stack (or anywhere else.)
+  //
+  // NOTE: This function only makes its precondition checks for non-constant
+  // inputs. If it makes them for constant inputs, it prevents the compiler from
+  // emitting equivalent compile-time diagnostics because the call to
+  // Builtin.stackAlloc() becomes unreachable.
+  if _isComputed(capacity) {
+    _precondition(capacity > 0, "Allocation capacity must be greater than or equal to zero")
+  }
+  let stride = MemoryLayout<T>.stride
+  let (byteCount, overflow) = capacity.multipliedReportingOverflow(by: stride)
+  if _isComputed(capacity) {
+    _precondition(!overflow, "Allocation byte count too large")
+  }
+  return byteCount
+}
+
+/// Will an allocation of the specified size fit on the stack?
+///
+/// - Parameters:
+///   - byteCount: The number of bytes to temporarily allocate. `byteCount` must
+///     not be negative.
+///   - alignment: The alignment of the temporary allocation. `alignment` must
+///     be a whole power of 2.
+///
+/// - Returns: Whether or not there is sufficient space on the stack to allocate
+///   `byteCount` bytes of memory.
+@_alwaysEmitIntoClient @_transparent
+internal func _isStackAllocationSafe(byteCount: Int, alignment: Int) -> Bool {
+  // PRECONDITIONS: Non-positive alignments are nonsensical, as are
+  // non-power-of-two alignments.
+  if _isComputed(alignment) {
+    _precondition(alignment > 0, "Alignment value must be greater than zero")
+    _precondition(_isPowerOf2(alignment), "Alignment value must be a power of two")
+  }
+
+  // If the alignment is larger than MaximumAlignment, the allocation is always
+  // performed on the heap. There are two reasons why:
+  // 1. llvm's alloca instruction can take any power-of-two alignment value, but
+  //    will produce unsafe assembly when that value is very large (i.e. it
+  //    risks a stack overflow.)
+  // 2. For non-constant values, we have no way to know what value to pass to
+  //    alloca and always pass MaximumAlignment. This may be incorrect if the
+  //    caller really wants a larger allocation.
+  if alignment > _minAllocationAlignment() {
+    return false
+  }
+
+  // Allocations smaller than this limit are reasonable to allocate on the stack
+  // without worrying about running out of space, and the compiler would emit
+  // such allocations on the stack anyway when they represent structures or
+  // stack-promoted objects.
+  if byteCount <= 1024 {
+    return true
+  }
+
+  // Finally, take a slow path through the standard library to see if the
+  // current environment can accept a larger stack allocation.
+  guard #available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *) else {
+    return false
+  }
+  return swift_stdlib_isStackAllocationSafe(byteCount, alignment)
```

Comment:
Just to double check: Is this symbol guaranteed to have weak linkage? IIUC we don't have availability for runtime entry points, and the preceding `guard #available` only protects against actually calling it; it doesn't affect linker behavior.